genson/src/test/java/com/owlike/genson/reflect/UnknownPropertyHandlerTest.java

```diff
@@ -0,0 +1,187 @@
+                "  \"spouse\":{\n" +
+                "    \"@class\":\"com.owlike.genson.reflect.UnknownPropertyHandlerTest$EvolvablePerson\",\n" +
+                "    \"age\":40,\n" +
+                "    \"name\":\"Marge\"\n" +
+                "  },\n" +
+                "  \"children\":[\n" +
+                "    \"Bart\",\n" +
+                "    \"Lisa\",\n" +
+                "    \"Maggie\"\n" +
+                "  ],\n" +
+                "  \"salary\":10000.0,\n" +
+                "  \"donutLover\":true\n" +
+                "}";
+
+        EvolvablePerson homer = GENSON.deserialize(json, EvolvablePerson.class);
+        assertEquals("Homer", homer.name);
+        assertEquals(50, homer.age);
+        assertEquals(Arrays.asList("Bart", "Lisa", "Maggie"), homer.unknownProperties.get("children"));
+        assertEquals(10_000d, homer.unknownProperties.get("salary"));
+        assertEquals(true, homer.unknownProperties.get("donutLover"));
+    }
+
+    @Test
+    public void testCtorDeserialization() {
+        String json = "{\n" +
+                "  \"@class\":\"com.owlike.genson.reflect.UnknownPropertyHandlerTest$CtorEvolvablePerson\",\n" +
+                "  \"age\":50,\n" +
+                "  \"name\":\"Homer\",\n" +
+                "  \"spouse\":{\n" +
+                "    \"@class\":\"com.owlike.genson.reflect.UnknownPropertyHandlerTest$CtorEvolvablePerson\",\n" +
+                "    \"age\":40,\n" +
+                "    \"name\":\"Marge\"\n" +
+                "  },\n" +
+                "  \"children\":[\n" +
+                "    \"Bart\",\n" +
+                "    \"Lisa\",\n" +
+                "    \"Maggie\"\n" +
+                "  ],\n" +
+                "  \"salary\":10000.0,\n" +
+                "  \"donutLover\":true\n" +
+                "}";
+
+        EvolvablePerson homer = GENSON.deserialize(json, CtorEvolvablePerson.class);
+        assertEquals("Homer", homer.name);
+        assertEquals(50, homer.age);
+        assertEquals(Arrays.asList("Bart", "Lisa", "Maggie"), homer.unknownProperties.get("children"));
+        assertEquals(10_000d, homer.unknownProperties.get("salary"));
+        assertEquals(true, homer.unknownProperties.get("donutLover"));
+    }
+
+    @Test
+    public void testRoundTrip() {
+        EvolvablePerson homer = new EvolvablePerson("Homer", 50);
+        homer.unknownProperties().put("spouse", new EvolvablePerson("Marge", 40));
+        homer.unknownProperties().put("children", Arrays.asList("Bart", "Lisa", "Maggie"));
+        homer.unknownProperties().put("salary", 10_000d);
+        homer.unknownProperties().put("donutLover", true);
+
+        String json = GENSON.serialize(homer);
+        EvolvablePerson homer2 = GENSON.deserialize(json, EvolvablePerson.class);
+
+        assertEquals(homer, homer2);
+    }
+
+    interface Evolvable {
+        Map<String, Object> unknownProperties();
+    }
+
+    static class EvolvableHandler implements UnknownPropertyHandler {
```

Comment:
Worth providing a simple impl like that as part of the lib.
In that case I think it might be better to not mutate the result of unknownProperties (for ex. this would then fail for an immutable map) and instead just define a addProperty method. Also some alternative namings could be PropertyBag, WithUnknownProperties, UnknownProperties.

Reply:
I (mostly) agree :-)

The reason I didn't do it is because any useful implementation must make certain assumptions about how the data class is implemented (ie. it must implement `Evolvable` and expose a holder map). If Java had mix-ins this would be a no-brainer, but unfortunately it doesn't, so it's really a question of whether we think an interface provides enough value to include it, and to create a default UPH implementation. If you feel we should, I'm not against it.

As for not mutating property map directly, it can get complicated. I originally had what you are suggesting, a pair of `addUnknownProperty/getUnknownProperties` methods. However, in that case you'd end up with the unknown properties serialized twice -- once as `unknownProperties` map, because of the getter, and once by the `BeanDescriptor` because they are, well, unknown properties.

In the end I decided that the simplest solution was to expose the mutable map via a single method that won't impact normal serialization. Now, the issue is still there -- if the field-based serialization was enabled, we'd have to mark the field itself with `@JsonIgnore` or make it transient. Again, a lot of this depends on personal preferences, and for test purposes I just did the easiest thing that worked. If we decide to add a standard implementation into Genson, we should probably discuss how it should look like and what the external contract and semantics should be.

Reply:
I think that a default impl would be useful if we were able to address the points you mention.
That way the user would only have to implement the contract without needing to also implement the handler and we could mitigate the double ser. problem.

1) To address the later problem we could provide an abstract class instead of an interface (mark the field as transient and annotate the methods). This would still be annoying for people that can't extend the class but would work in every case for all the others.

OR/AND (we can do both)

2) Have an interface with JsonIgnore (or use names that don't match std get/set patterns like you did in this test) and as a bonus fix this issue https://github.com/owlike/genson/issues/85 to merge JsonIgnore with subclasses. 

Fixing this issue is not trivial, so that could be done in a second time later on. The main problem is due to the fact that exclusion of properties happens before merging of annotations on all discovered properties. This is an useful feature to have...

OR

3) Do something like this hacky idea from https://stackoverflow.com/a/25907755/840441. Though this is probably too hacky for what we need :)

Reply:
We can certainly do 1) and 2) and provide both an interface and an abstract base class that implements it. 

I'm not a big fan of base classes, as most Java developers are not happy when you force them to extend something and take away their only opportunity for inheritance, but I guess providing it as an option and not something that's required softens the blow a bit -- they can always choose not to use our base class and implement the interface directly in their own class(es).

3) scares me... ;-)